docs/tools-custom.qmd

```diff
@@ -97,6 +97,101 @@ The following instance methods are available to tools that need to interact with
 
 See the documentation on [Sandbox Environments](sandboxing.qmd) for additional details.
 
+## Stateful Tools
+
+Some tools need to retain state across invocations (for example, the `bash_session()` and `web_browser()` tools both interact with a stateful remote process). You can create stateful tools by using the `store_as()` function to access discrete storage for your tool and/or speicfic instances of your tool.
+
+For example, imagine we were creating a `web_surfer()` tool that builds  on the `web_browser()` tool to complete sequences of browser actions in service of researching a topic. We might want to ask multiple questions of the web surfer and have it retain its message history and browser state.
+
+Here's the complete source code for this tool. 
+
+```python
+from textwrap import dedent
+
+from pydantic import Field
+from shortuuid import uuid
+
+from inspect_ai.model import (
+  ChatMessage, ChatMessageSystem, ChatMessageUser, get_model
+)
+from inspect_ai.tool import Tool, tool, web_browser
+from inspect_ai.util import StoreModel, store_as
+
+class WebSurferState(StoreModel):
+    messages: list[ChatMessage] = Field(default_factory=list)
+
+@tool
+def web_surfer(instance: str | None = uuid()) -> Tool:
```

Comment:
I'm on my first pass reading through this, but it seems like you've chosen to expose the previously _internal_ `session_name` concept implemented by stateful tools in the form of `instance`. Is my understanding correct?

Reply:
It's a store level concept which can be used flexibly (e.g. multiple agents or tools could share a store instance e.g. for a multi-player game). It maps nicely to `session_name` though in the case of your stateful backend.